File: py/636.exclusive-time-of-functions.py
#
# @lc app=leetcode id=636 lang=python3
#
# [636] Exclusive Time of Functions
#
from typing import *
# @lc code=start


class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        tasks_stack = []
        tasks_exc_time = [0]*n
        run_time = 0
        for log in logs:
            task_num = int(log[0:log.find(':')])
            time_num = int(log[log.rfind(':')+1:])
            status = log[log.find(':')+1:log.rfind(':')]
            if status == "start":
                if len(tasks_stack):
                    tasks_exc_time[tasks_stack[-1]] += time_num-run_time
                tasks_stack.append(task_num)
                run_time = time_num
            else:
                tasks_exc_time[task_num] += time_num - run_time + 1
                tasks_stack.pop()
                run_time = time_num + 1
        return tasks_exc_time
# @lc code=end

File: 636.exclusive-time-of-functions.cpp
/*
 * @lc app=leetcode id=636 lang=cpp
 *
 * [636] Exclusive Time of Functions
 */
#include <vector>
#include <string>
using namespace std;
// @lc code=start
class Solution
{
public:
    vector<int> exclusiveTime(int n, vector<string> &logs)
    {
        vector<int> tasks_stack;
        vector<int> tasks_exc_time(n);
        int run_time = 0;
        for (auto &log : logs)
        {
            int task_num = stoi(log.substr(0, log.find_first_of(':')));
            int time_num = stoi(log.substr(log.find_last_of(':') + 1));
            string status = log.substr(log.find_first_of(':') + 1, log.find_last_of(':') - log.find_first_of(':') - 1);
            if (status == "start")
            {
                if (!tasks_stack.empty())
                {
                    tasks_exc_time[tasks_stack.back()] += time_num - run_time;
                }
                tasks_stack.push_back(task_num);
                run_time = time_num;
            }
            else
            {
                tasks_exc_time[task_num] += time_num - run_time + 1;
                tasks_stack.pop_back();
                run_time = time_num + 1;
            }
        }
        return tasks_exc_time;
    }
};
// @lc code=end
